More changes toward an end-to-end flow with the new APIS

chavdar opened the pull request:
Changes to the new API:
- Made GobblinInstanceLauncher extend configurable since it is responsible of holding the system configuration
- Added initial version of JobExecutionDriver and modified JobExecutionLauncher to return JobExecutionDriver
- Added a JobLauncherExecutionDriver which acts as an adapter to the legacy JobLauncher/JobLauncherFactory API
- Reworked JobSpec.Builder to be convention-based and be able to reuse some of the properties in the Job configuration
- Added StandardGobblinInstanceDriver and StandardGobblinInstanceLauncher-- a very rudimentary implementations of GobblinInstanceDriver and GobblinInstanceLauncher

Changes to legacy code:
- Documentation and helper methods in JobState to decrease reliance on direct access to job properties
- Helper methods in JobContext
- Added a method JobLauncherFactory. newJobLauncher() method to allow instantiating a specific JobLauncher


@ibuenros @autumnust can you review ?

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/api/JobExecutionState.java:
```diff
@@ -36,4 +54,86 @@ public JobExecutionState(JobSpec jobSpec, JobExecution jobExecution) {
+  }
+
+  public void switchToPending() {
+    Preconditions.checkState(null == this.runningState, "unexpected state:" + this.runningState);
```
nit: `String.format("unexpected stage transition: %s to %s", this.runningState, PENDING)`, also for the other similar methods.

chavdar replied:
Will fix.

chavdar commented on gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java:
```diff
@@ -137,7 +137,8 @@ public JobSpecRunnable(JobSpec jobSpec) {
     public void run() {
-       _jobLauncher.launchJob(_jobSpec, new JobExecutionListener());
+       _jobLauncher.launchJob(_jobSpec);
+       // new JobExecutionListener();
```
This is incomplete and will get changed somewhat in a following commit;

chavdar commented:
Since a lot of changes have accumulated, I'll re-summarize the changes:

Changes to the new API:

* GobblinInstanceLauncher/GobblinInstanceDriver
    - Renamed GobblinInstanceLauncher.getName() to GobblinInstanceLauncher.getInstanceName()
    - Made GobblinInstanceLauncher extend configurable since it is responsible of holding the system configuration

* JobCatalog/JobSpec:
    - Refactored the register/unregister methods into a separate interface - JobCatalogListenersContainer
    - Reworked JobSpec.Builder to be convention-based and be able to reuse some of the properties in the Job configuration

*  JobExecutionLauncher/JobExecutionDriver
    - Added the class JobExecutionDriver to control/monitor the execution of a single gobblin job
    - JobExecutionLauncher is not a factory for JobExecutionDriver instances
    - Add JobExecutionStateListenerContainer interface for managing JobExecutionStateListener instances

* JobExecutionStatus/JobExecutionState
    - JobExecutionStatus is now an interface
    - JobExecutionState is now responsible for the state management (including callbacks)

Implementations:

- Added a sample implementation JobLauncherExecutionDriver of JobExecutionDriver which acts as an adapter to the legacy JobLauncher/JobLauncherFactory API
- Added DefaultGobblinInstanceDriverImpl which accepts all components but it is not responsible for managing them. This implementation can be used as a shell to implement a typical Gobblin job lifecycle
- Added StandardGobblinInstanceDriver which will be responsible for instantiating and managing the instance components and passing them to DefaultGobblinInstanceDriverImpl
- Added StandardGobblinInstanceLauncher-- a very rudimentary implementation to instantiate StandardGobblinInstanceDriver
- Standard implementations for JobCatalogListenersContainer (JobCatalogListenersList) and JobExecutionStateListenerContainer (JobExecutionStateListeners)
- The above use a generic callback infrastructure in gobblin-utility/gobblin/util/callbacks
  - Callbacks a typically represented through a Function<L, R> where L is the listener type and R is the callback result (or Void)
  - There is a helper class Callback<L, R> to let adding some human-readable description of the callback
  - Callbacks are created by a CallbackFactory which is responsble for binding any parameters that need to be passed to the callback
  - Callbacks are managed by CallbacksDispatcher

Changes to legacy code:

- Documentation and helper methods in JobState to decrease reliance on direct access to job properties
- Added a description of current state diagram for RunningState based on code inspection
- Helper methods in JobContext
- Added a method JobLauncherFactory. newJobLauncher() method to allow instantiating a specific JobLauncher type instead of relying on the job config

chavdar commented:
As discussed, there some follow-up items:
- Have an internal implementation of JobExecutionStatus in JobExecutionState that can be easily serialized/deserialized
- Add more javadocs to CallbacksDispatcher to explain usage

I'll make the above in a separate PR since my local copy is already ahead of this PR and I'd rather don't merge any more changes in this PR.

ibuenros commented on gobblin-utility/src/main/java/gobblin/util/callbacks/CallbacksDispatcher.java:
```diff
@@ -0,0 +1,160 @@
+    return execCallbacks(callbackFactory, listeners);
+  }
+
+  private <R> CallbackResults<L, R> execCallbacks(CallbackFactory<L, R> callbackFactory, List<L> listeners)
```
Couldn't you pass the `Callback` directly?

chavdar replied:
Yeah, fixed.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/instance/DefaultGobblinInstanceDriverImpl.java:
```diff
@@ -137,7 +137,8 @@ public JobSpecRunnable(JobSpec jobSpec) {
 
     @Override
     public void run() {
```
Run is generally not used in an `AbstractIdleService`, should you use a `AbstractExecutionThreadService` instead?

chavdar replied:
This is actually a method of JobSpecRunnable which gets passed to the JobExecutionScheduler.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java:
```diff
@@ -1,154 +1,148 @@
-      callbackMsg = "callback " + callback + " on " + listener;
-      _log.debug("Started: " + callbackMsg);
-    }
+  public void callbackOneListener(Function<JobCatalogListener, Void> callback,
```
What does this method do?

chavdar replied:
This is needed for the contract, that when we add a new listener, it should get callbacks for all existing jobs. See InMemoryJobCatalog.addListener() .

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java:
```diff
@@ -1,154 +1,148 @@
-      sb.append(']');
-      sb.append(')');
-      return sb.toString();
+  public static class AddJobCallback extends Callback<JobCatalogListener, Void> {
```
Could this be inner static classes of `JobCatalogListener` instead? They are not specific to this class.

chavdar replied:
Sure. Makes sense.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java:
```diff
@@ -0,0 +1,191 @@
+      }
+    }
+
+    _legacyLauncher.close();
```
Will this prevent the job from running?

chavdar replied:
Yes, this shut downs the ExecutorService in AbstractJobLauncher.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java:
```diff
@@ -0,0 +1,191 @@
+    public void onJobPrepare(JobContext jobContext) throws Exception {
+      super.onJobPrepare(jobContext);
+      _jobContext = jobContext;
+      _jobState.switchToPending();;
```
extra `;`

chavdar replied:
Fixed.

ibuenros commented on gobblin-runtime/src/main/java/gobblin/runtime/job_exec/JobLauncherExecutionDriver.java:
```diff
@@ -0,0 +1,191 @@
+    _legacyLauncher.close();
+  }
+
+  class MyJobListener extends AbstractJobListener {
```
Can you change the name of the class? Maybe `JobListenerToJobStateBridge` or something more meaningful?

chavdar replied:
Done.

ibuenros commented on gobblin-runtime/src/test/java/gobblin/runtime/std/TestJobExecutionStateListeners.java:
```diff
@@ -0,0 +1,77 @@
+public class TestJobExecutionStateListeners {
+
+  @Test
+  public void testHappyPath() {
```
Interesting name :P
